File: src/main/other/advent_of_code_2022/day4/src/part1.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

use itertools::Itertools;

use crate::interval::Interval;

pub fn part1() -> Result<u32, Box<dyn std::error::Error>> {
    let file = File::open("input/input.txt")?;
    let reader = BufReader::new(file);

    let mut res = 0;
    for line in reader.lines() {
        let line = line.unwrap();
        let (first, second) = line
            .split(',')
            .map(|str| Interval::from_string(str))
            .next_tuple()
            .unwrap();

        if first.contains_fully(&second) || second.contains_fully(&first) {
            res += 1
        }
    }

    Ok(res)
}

File: src/main/other/advent_of_code_2023/day4/part1.rb
input = File.open(File.join(File.dirname(__FILE__), "input.txt"), "r").map(&:chomp)

res = input.sum do |line|
    line.split('|').map { |s| s.scan(/\d+/).map(&:to_i) }.then.sum do |first, second|
        (2 ** ((first[1...] & second).size - 1)).floor
    end
end

pp res
